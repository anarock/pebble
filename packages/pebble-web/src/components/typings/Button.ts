import * as React from "react";
import * as CSS from "csstype";
import { Interpolation } from "./emotionCustom";

export type ButtonType = "primary" | "secondary" | "link" | "success" | "alert";

export interface ButtonProps {
  children: React.ReactNode;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
  showShadow?: boolean;
  disabled?: boolean;
  type?: ButtonType;
  size?: "x-small" | "small" | "large";
  width?: CSS.WidthProperty<number>;
  styles?: Interpolation;
  showRipple?: boolean;
  loading?: boolean;
  filled?: boolean;
  buttonProps?: React.ButtonHTMLAttributes<HTMLButtonElement>;
}

export interface DropDownButtonProps extends ButtonProps {
  isOpen: boolean;
  isSelected: boolean;
}

interface MappingColorByTypeOptions {
  textColor?: string;
  base: string;
  hover: string;
  active: string;
  disabled: string;
}

export interface MappingColorByType {
  primary: MappingColorByTypeOptions;
  secondary: MappingColorByTypeOptions;
  success: MappingColorByTypeOptions;
  alert: MappingColorByTypeOptions;
}
