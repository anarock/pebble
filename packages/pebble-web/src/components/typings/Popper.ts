import * as React from "react";
import { PopperProps as PopperProps_ } from "react-popper";
import { Interpolation } from "./emotionCustom";

type Label =
  | string
  | number
  | ((args: { toggle: () => void; isOpen: boolean }) => React.ReactNode);

export interface PopperProps
  extends Omit<PopperProps_, "positionFixed" | "children"> {
  label: Label;
  popperBackgroundColor?: string;
  children: (args: { toggle: () => void; isOpen: boolean }) => React.ReactNode;
  isOpen?: boolean;
  controlled?: boolean;
  popperStyles?: Interpolation;
  closeOnOutsideClick: boolean;
  onOutsideClick?: () => void;
}

export interface PopperState {
  isOpen: boolean;
}
