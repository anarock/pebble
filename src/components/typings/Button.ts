import * as React from "react";

export type ButtonType = "primary" | "secondary" | "link" | "success" | "alert";

type Children = JSX.Element | string | number | React.ReactChildren;

export interface ButtonProps {
  children: Children;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
  showShadow?: boolean;
  disabled?: boolean;
  type?: ButtonType;
  size?: "x-small" | "small" | "large";
  width?: number;
  className?: string;
  showRipple?: boolean;
  loading?: boolean;
  filled?: boolean;
}

export interface DropDownButtonProps extends ButtonProps {
  isOpen: boolean;
  isSelected: boolean;
}

interface MappingColorByTypeOptions {
  textColor?: string;
  base: string;
  hover: string;
  active: string;
  disabled: string;
}

export interface MappingColorByType {
  primary: MappingColorByTypeOptions;
  secondary: MappingColorByTypeOptions;
  success: MappingColorByTypeOptions;
  alert: MappingColorByTypeOptions;
}
