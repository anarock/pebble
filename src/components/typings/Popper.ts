import * as React from "react";
import { PopperProps as PopperProps_ } from "react-popper";

// @ts-ignore
export interface PopperProps extends Partial<PopperProps_> {
  label:
    | string
    | number
    | ((args?: { toggle: () => void; isOpen: boolean }) => JSX.Element);
  popperBackgroundColor?: string;
  children: (
    args: { toggle: () => void; isOpen: boolean }
  ) => React.ReactChildren | React.ReactChild;
  isOpen?: boolean;
  controlled?: boolean;
}

export interface PopperState {
  isOpen: boolean;
}
