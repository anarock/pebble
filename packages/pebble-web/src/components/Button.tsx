import * as React from "react";

import {
  iconStyle,
  getButtonStyle,
  dropDownButtonStyle,
  dropDownButtonDefaultStyle
} from "./styles/Button.styles";
import Ink from "react-ink";
import { ButtonProps, DropDownButtonProps } from "./typings/Button";
import Loader from "./Loader";
import { colors } from "pebble-shared";

const Button: React.FunctionComponent<ButtonProps> = ({
  type = "primary",
  disabled,
  children,
  onClick,
  width,
  showShadow,
  styles,
  showRipple = true,
  loading,
  size = "small",
  buttonProps
}: ButtonProps) => {
  const disableAction = disabled || loading;

  const filled = size !== "x-small";

  return (
    <button
      css={[getButtonStyle(size, type, !!showShadow, filled), styles]}
      onClick={!disableAction ? onClick : undefined}
      style={{ width }}
      disabled={disabled}
      {...buttonProps}
    >
      {loading ? <Loader color={colors.white.base} scale={0.4} /> : children}
      {!disableAction && showRipple && type !== "link" && <Ink />}
    </button>
  );
};

export const DropDownButton = ({
  isOpen,
  isSelected,
  children,
  styles,
  ...props
}: DropDownButtonProps) => {
  return (
    <Button
      {...props}
      type="secondary"
      styles={[
        dropDownButtonStyle,
        !(isOpen || isSelected) && dropDownButtonDefaultStyle,
        styles
      ]}
    >
      {children}{" "}
      <i
        className="pi pi-arrow-drop-down"
        css={iconStyle}
        style={{
          transform: isOpen ? "rotate(180deg)" : "none"
        }}
      />
    </Button>
  );
};

export default Button;
