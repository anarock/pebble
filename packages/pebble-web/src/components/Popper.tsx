import * as React from "react";
import { PopperProps, PopperState } from "./typings/Popper";
import { Manager, Reference, Popper } from "react-popper";
import { arrowStyle, popperStyle } from "./styles/Popper.styles";
import { colors } from "pebble-shared";

import OutsideClick from "./OutsideClick";
import MountTransition from "./shared/MountTransition";

export default class PebblePopper extends React.PureComponent<
  PopperProps,
  PopperState
> {
  static defaultProps: Partial<PopperProps> = {
    placement: "bottom",
    closeOnOutsideClick: true
  };

  state: PopperState = {
    isOpen: !!this.props.isOpen
  };

  private toggle = () => {
    this.setState({
      isOpen: !this.state.isOpen
    });
  };

  render() {
    const {
      label,
      popperBackgroundColor = colors.white.base,
      children,
      controlled,
      isOpen,
      popperStyles,
      onOutsideClick,
      ...props
    } = this.props;

    const _isPopperOpen = controlled ? !!isOpen : this.state.isOpen;

    return (
      <OutsideClick
        onOutsideClick={() => {
          this.setState({
            isOpen: false
          });
          if (onOutsideClick) {
            onOutsideClick();
          }
        }}
        disabled={!_isPopperOpen}
      >
        <Manager>
          <Reference>
            {({ ref }) => (
              <div style={{ display: "inline-block" }} ref={ref}>
                {typeof label === "function"
                  ? label({ toggle: this.toggle, isOpen: this.state.isOpen })
                  : label}
              </div>
            )}
          </Reference>

          <MountTransition visible={_isPopperOpen}>
            {transitionStyles => (
              <Popper {...props} positionFixed>
                {({ ref, style, placement, arrowProps }) => {
                  const wrapperStyle = {
                    ...style,
                    ...transitionStyles,
                    backgroundColor: popperBackgroundColor,
                    transform: `${style.transform ||
                      ""} ${transitionStyles.transform || ""}`,
                    transformOrigin: `${arrowProps.style.left ||
                      0}px ${arrowProps.style.top || 0}px`
                  };

                  return (
                    <div
                      css={[popperStyle, wrapperStyle]}
                      ref={ref}
                      data-placement={placement}
                    >
                      {children({
                        toggle: this.toggle,
                        isOpen: this.state.isOpen
                      })}
                      <div
                        css={arrowStyle}
                        ref={arrowProps.ref}
                        style={{
                          ...arrowProps.style,
                          color: popperBackgroundColor
                        }}
                        data-placement={placement}
                      >
                        â–¶
                      </div>
                    </div>
                  );
                }}
              </Popper>
            )}
          </MountTransition>
        </Manager>
      </OutsideClick>
    );
  }
}
