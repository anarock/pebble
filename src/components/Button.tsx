import * as React from "react";
import { cx } from "emotion";
import {
  iconStyle,
  getButtonStyle,
  dropDownButtonStyle,
  dropDownButtonDefaultStyle
} from "./styles/Button.styles";
import Ink from "react-ink";
import { ButtonProps, DropDownButtonProps } from "./typings/Button";
import Loader from "./Loader";
import { colors } from "../theme";

const Button: React.FunctionComponent<ButtonProps> = ({
  type = "primary",
  disabled,
  children,
  onClick,
  width,
  showShadow,
  className,
  showRipple = true,
  loading,
  size = "small"
}: ButtonProps) => {
  const disableAction = disabled || loading;

  const filled = size !== "x-small";
  const _className = cx(
    getButtonStyle(size, type, !!showShadow, filled),
    className
  );

  return (
    <button
      className={_className}
      onClick={!disableAction ? onClick : undefined}
      style={{ width }}
      disabled={disabled}
    >
      {loading ? <Loader color={colors.white.base} scale={0.4} /> : children}
      {!disableAction && showRipple && type !== "link" && <Ink />}
    </button>
  );
};

export const DropDownButton: React.FunctionComponent<DropDownButtonProps> = ({
  isOpen,
  isSelected,
  children,
  className,
  ...props
}) => {
  const _className = cx(dropDownButtonStyle, {
    [dropDownButtonDefaultStyle]: !(isOpen || isSelected)
  });

  return (
    <Button {...props} type="secondary" className={cx(_className, className)}>
      <React.Fragment>
        {children}{" "}
        <i
          className={cx("pi pi-arrow-drop-down", iconStyle)}
          style={{
            transform: isOpen ? "rotate(180deg)" : "none"
          }}
        />
      </React.Fragment>
    </Button>
  );
};

export default Button;
