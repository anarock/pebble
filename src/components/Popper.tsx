import * as React from "react";
import { PopperProps, PopperState } from "./typings/Popper";
import { Manager, Reference, Popper } from "react-popper";
import { arrowStyle, popperStyle } from "./styles/Popper.styles";
import { colors } from "../theme";
import { cx } from "react-emotion";
import { Transition } from "react-spring";
import OutsideClick from "./OutsideClick";

export default class extends React.PureComponent<PopperProps, PopperState> {
  static defaultProps: Partial<PopperProps> = {
    placement: "bottom",
    popperBackgroundColor: colors.white.base,
    closeOnClickOutside: true
  };

  state: PopperState = {
    isOpen: this.props.isOpen
  };

  private toggle = () => {
    this.setState({
      isOpen: !this.state.isOpen
    });
  };

  render() {
    const {
      label,
      popperBackgroundColor,
      children,
      controlled,
      isOpen,
      popperClassName,
      ...props
    } = this.props;

    const _isPopperOpen = controlled ? isOpen : this.state.isOpen;

    return (
      <OutsideClick
        onOutsideClick={() =>
          this.setState({
            isOpen: false
          })
        }
        disabled={!_isPopperOpen}
      >
        <Manager>
          <Reference>
            {({ ref }) => (
              <div ref={ref}>
                {typeof label === "function"
                  ? label({ toggle: this.toggle, isOpen: this.state.isOpen })
                  : label}
              </div>
            )}
          </Reference>
          <Transition
            from={{ opacity: 0 }}
            enter={{ opacity: 1 }}
            leave={{ opacity: 0 }}
          >
            {_isPopperOpen &&
              (styles => (
                <Popper {...props}>
                  {({ ref, style, placement, arrowProps }) => (
                    <div
                      className={cx(popperStyle, popperClassName)}
                      ref={ref}
                      style={{
                        ...styles,
                        ...style,
                        backgroundColor: popperBackgroundColor
                      }}
                      data-placement={placement}
                    >
                      {children({
                        toggle: this.toggle,
                        isOpen: this.state.isOpen
                      })}
                      <div
                        className={arrowStyle}
                        ref={arrowProps.ref}
                        style={{
                          ...arrowProps.style,
                          color: popperBackgroundColor
                        }}
                        data-placement={placement}
                      >
                        â–¶
                      </div>
                    </div>
                  )}
                </Popper>
              ))}
          </Transition>
        </Manager>
      </OutsideClick>
    );
  }
}
