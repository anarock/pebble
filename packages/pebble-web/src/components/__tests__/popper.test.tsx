import * as React from "react";
import renderer from "react-test-renderer";
import Button from "../Button";
import Popper from "../Popper";
// import { mount } from "enzyme";

describe("Component: Popper", () => {
  test("snapshot", () => {
    const component = renderer.create(
      <Popper
        label={({ toggle }) => <Button onClick={toggle}>Click Me</Button>}
        placement="left"
        isOpen
      >
        {() => <div className="dropdown-content" />}
      </Popper>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Popper: Functionality", () => {
  // enzyme doesn't support React Context
  // test("click opens popper", () => {
  // 	const component = mount(<Popper
  // 		label={({ toggle }) => <Button className="button" onClick={toggle}>Click Me</Button>}
  // 		placement="left"
  // 		isOpen
  // 	>
  // 		{() => <div className="dropdown-content" />}
  // 	</Popper>);
  //
  // 	expect(component.find(".dropdown-content")).toHaveLength(0);
  // 	component.find("button").simulate("click");
  // 	expect(component.find(".dropdown-content")).toHaveLength(1);
  // })
});
