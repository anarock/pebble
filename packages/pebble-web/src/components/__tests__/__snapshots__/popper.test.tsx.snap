// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Popper snapshot 1`] = `
.emotion-0 {
  line-height: 23px;
  height: 40px;
  padding: 0 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all ease-out .2s;
  transition: all ease-out .2s;
  border-radius: 3px;
  outline: none;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  white-space: nowrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0;
  height: 40px;
  min-width: 100px;
  font-weight: 400;
  color: #ffffff;
  font-size: 12px;
  background-color: #6161FF;
  border: none;
  box-shadow: none;
}

.emotion-0[disabled] {
  cursor: not-allowed;
}

.emotion-0:not([disabled]):hover {
  color: #ffffff;
  background-color: #7070FF;
  border-color: #7070FF;
}

.emotion-0:not([disabled]):active {
  color: #ffffff;
  background-color: #5656ED;
  border-color: #5656ED;
}

.emotion-0[disabled] {
  color: #ffffff;
  background-color: #C0C0FF;
  border-color: #C0C0FF;
}

.emotion-2 {
  margin: 14px;
  box-shadow: 0 2px 15px 0 rgba(0,0,0,0.1);
  border-radius: 3px;
  position: absolute;
  top: 0;
  left: 0;
  opacity: 0;
  pointer-events: none;
  -webkit-transform: scale(0.95);
  -ms-transform: scale(0.95);
  transform: scale(0.95);
  background-color: #ffffff;
  -webkit-transform-origin: 0px 0px;
  -ms-transform-origin: 0px 0px;
  transform-origin: 0px 0px;
}

.emotion-1 {
  position: absolute;
  text-shadow: 1px 0 20px rgba(0,0,0,0.1);
}

.emotion-1[data-placement^='top'] {
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
  bottom: -11px;
}

.emotion-1[data-placement^='bottom'] {
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  top: -11px;
}

.emotion-1[data-placement^='right'] {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
  left: -11px;
}

.emotion-1[data-placement^='left'] {
  right: -11px;
}

<div
  className="emotion-3"
>
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <button
      className="emotion-0"
      onClick={[Function]}
      style={
        Object {
          "width": undefined,
        }
      }
    >
      Click Me
      <canvas
        height={0}
        onDragOver={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "borderRadius": "inherit",
            "height": "100%",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
        width={0}
      />
    </button>
  </div>
  <div
    className="emotion-2"
  >
    <div
      className="dropdown-content"
    />
    <div
      className="emotion-1"
      style={
        Object {
          "color": "#ffffff",
        }
      }
    >
      â–¶
    </div>
  </div>
</div>
`;
